testsuite/tests/typing-layouts-unboxed-records/unique.ml

```diff
@@ -0,0 +1,92 @@
+(* TEST
+ flambda2;
+ include stdlib_upstream_compatible;
+ flags = "-extension layouts_alpha -extension unique";
+ {
+   expect;
+ }
+*)
+
+(* Uniqueness tests *)
+
+type t = #{ x : string ; y : string }
+let use_string (_s @ unique) = ()
+let mk : unit -> t @ unique = fun () -> #{ x = "hi"; y = "hi" }
+[%%expect{|
+type t = #{ x : string; y : string; }
+val use_string : ('a : value_or_null). 'a @ unique -> unit = <fun>
+val mk : unit -> t @ unique = <fun>
+|}]
+
+(* Can access different fields *)
+let () =
+  let t = mk () in
+  use_string t.#x;
+  use_string t.#y
+[%%expect{|
+|}]
+
+let () =
+  let #{ x ; y } = mk () in
+  use_string x;
+  use_string y
+[%%expect{|
+|}]
+
+(* Cannot access the same field twice *)
+let () =
+  let t = mk () in
+  use_string t.#x;
+  use_string t.#x
+[%%expect{|
+Line 4, characters 13-17:
+4 |   use_string t.#x
+                 ^^^^
+Error: This value is used here, but it has already been used as unique:
+Line 3, characters 13-17:
+3 |   use_string t.#x;
+                 ^^^^
+
+|}]
+
+let () =
+  let #{ x ; y = _ } = mk () in
+  use_string x;
+  use_string x
+[%%expect{|
+Line 4, characters 13-14:
+4 |   use_string x
+                 ^
+Error: This value is used here, but it has already been used as unique:
+Line 3, characters 13-14:
+3 |   use_string x;
+                 ^
+
+|}]
+
+(* A functional update to a field allows a field to be reused *)
+let () =
+  let t = mk () in
+  use_string t.#x;
+  let t = #{ t with x = "fresh" } in
+  use_string t.#x
+[%%expect{|
+|}]
+
+(* But not a functional update to a different field *)
+let () =
+  let t = mk () in
+  use_string t.#x;
+  let t = #{ t with y = "fresh" } in
+  use_string t.#x
+[%%expect{|
+Line 4, characters 13-14:
+4 |   let t = #{ t with y = "fresh" } in
+                 ^
+Error: This value is used here, but it has already been used as unique:
+Line 3, characters 13-17:
+3 |   use_string t.#x;
+                 ^^^^
+
+|}]
+
```

Comment:
Can you add the following tests:
```
let () =
  let t = mk () in
  unique_use t;
  let t = #{ t with y = "fresh" } in
  unique_use t

let () =
  let t = mk () in
  unique_use t;
  let t = #{ t with x = "fresh"; y = "fresh" } in
  unique_use t.#x

type t = #{ x : unit; y : string }
val mk : unit -> t @ unique = <fun>

let () =
  let t = mk () in
  unique_use t;
  let t = #{ t with y = "fresh" } in
  unique_use t

let () =
  let t = mk () in
  unique_use t;
  let t = #{ t with x = (); y = "fresh" } in
  unique_use t.#x
```

I would guess that the output will be fail, success, success, success (this assumes that `unit` mode-crosses uniqueness which it might only do with same extra type annotations).

In contrast, normal records will fail all four tests. This tests the fact that we implicitly borrow the memory address of records for the functional update while we do not do that for unboxed records.

Reply:
Thanks very much for the tests - I added the first, second, and fourth.

I couldn't find a way to annotate the third test such that it passes. It seems like we want to annotate the unit component "produced" by the functional update as unique, but there's no piece of syntax in the functional update expression that refers. Is there a trick I'm missing?

Reply:
Thanks, I agree now that the third test will fail; this is due to how mode crossing is implemented in the uniqueness analysis and while we should fix it at some point in the future, this is not the place to do so. Can you still add the test with the error message and add the following CR?
```
(* CR uniqueness: this test should succeed since unboxed records have no memory address
    and the first field is of type unit and thus mode-crosses uniqueness. 
    We should fix this by allowing multiple unique uses for values that mode-cross uniqueness. *)
```